### PR TITLE
Add new optional socrata features for running inventory log

### PR DESCRIPTION
**Pull request by johnclary**
In service of https://github.com/cityofaustin/atd-data-tech/issues/7581.

### Background

This issue ☝️ is asking for a daily snapshot of a Knack view of inventory items. The approach I've settled on is to simply append rows to a Socrata dataset on a daily basis, and pair that with a [PowerBI dashboard](https://github.com/cityofaustin/atd-data-tech/issues/8212) to interact with the data.

This can be accomplished purely through the Socrata dataset configuration, by **not** setting a [row identifier](https://support.socrata.com/hc/en-us/articles/360008065493) on the dataset. When no row identifier is configured, any attempt to upsert to the dataset will result in an insert, thereby appending rows to the dataset.

So we can use atd-knack-services to accomplish this, with a few minor enhancements:

1. For reporting purposes, we need a way to distinguish duplicate rows by the date of their snapshot. And so I've added an optional container config property, `append_timestamps_socrata`, which appends a timestamp value to each record before publishing to Socrata.

2. Because we're not storing these snapshots anywhere other than the dataset itself, there is a risk that this dataset could get accidentally overwritten and the historical record would be lost. Eek. I have taken a bit of shortcut here by adding an optional container config property, `no_replace_socrata`, which prevents a `replace` operation on the destination dataset. (W/o this change, a replace action is automatically triggered when ever `records_to_socrata.py` is run without a `--date` argument.)

I don't feel super awesome about choice `#2`. A more resilient solution would be to set up a proper pipeline where snapshots of the Knack view are archived in S3, and then piped into Socrata. That's a fairly well-worn pattern, but we're talking about a whole separate set of tooling to manage this one report :/ 

The reason I'm ok with this current approach is that the report—the quantity of inventory on hand, per item, on an arbitrary day—can be re-constructed through the inventory system's transaction log fairly easily. So on the off chance that this Socrata dataset were corrupted, we could rebuild it.

Thanks for reading this. I'd appreciate your feedback!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -214,7 +216,8 @@ $ python records_to_socrata.py \
 
 - `--app-name, -a` (`str`, required): the name of the source Knack application
 - `--container, -c` (`str`, required): the object or view key of the source container
-- `--date, -d` (`str`, optional): an ISO-8601-compliant date string. If no timezone is provided, GMT is assumed. Only records which were modified at or after this date will be processed. If excluded, all records will be processed.
+- `--date, -d` (`str`, optional): an ISO-8601-compliant date string. If no timezone is provided, GMT is assumed. Only records which were modified at or after this date will be processed. If excluded, all records will be processed and the destination dataset will be
+*completely replaced*.
```
Comment (johnclary) [inline README.md:220]: > If excluded, all records will be processed and the destination dataset will be
*completely replaced*.

This behavior has not changed; it just wasn't explicitly documented

File: README.md [hunk 2]
```diff
@@ -243,7 +246,8 @@ $ python records_to_agol.py \
...
```
Comment (johnclary) [inline README.md:250]: > If excluded, all records will be processed and the destination dataset will be
*completely replaced*.

This behavior has not changed; it just wasn't explicitly documented

File: services/config/knack.py [hunk 3]
```diff
@@ -134,14 +137,14 @@
...
-            "item_type": "layer"
+            "item_type": "layer",
         },
         "view_3003": {
             "description": "Signal detection status logs",
             "scene": "scene_514",
             "modified_date_field": "field_2565",
-            "socrata_resource_id": "e4b6-xseb"
-        }
+            "socrata_resource_id": "e4b6-xseb",
```
Comment (frankhereford) [inline services/config/knack.py:146]: I appreciate the `,`s after the last key-value pairs. Thanks for adding those.

File: services/utils/socrata.py [hunk 4]
```diff
@@ -1,5 +1,7 @@
+from datetime import datetime
 import os
 
+import arrow
```
Comment (frankhereford) [inline services/utils/socrata.py:4]: > Arrow is named after the arrow of time and is heavily inspired by moment.js and requests.

Inspired by moment is a +1 in my book. This is a new library to me - TIL.

File: services/utils/socrata.py [hunk 5]
```diff
@@ -21,6 +23,25 @@ def get_floating_timestamp_fields(resource_id, metadata):
...
+
+    Args:
+        records (list): A list of record dicts
+        key (str): The key which will be added to each record dict with the timestamp value
+        format_ (str): The arrow formatting token string 
+
+    Returns:
+        None: the records are updated in place.
+    """
+    now = arrow.now().to(tzinfo).format(format_)
```
Comment (frankhereford) [inline services/utils/socrata.py:40]: Magic. https://xkcd.com/1883/

**Review by frankhereford — APPROVED** [review]
I've gone over the code, and it looks great. I don't have a good local test environment setup currently to put this through its paces, so please take my approval with a grain of salt. That said, I don't see anything that should hold this up from merging. Thanks for writing this!

**Review by chiaberry — APPROVED** [review]

File: services/config/knack.py (hunk 3 above)
Comment (johnclary) [reply services/config/knack.py:146]: 👍 linted via [Black](https://github.com/psf/black)

File: services/utils/socrata.py (hunk 4 above)
Comment (johnclary) [reply services/utils/socrata.py:4]: 👍 It's not the most performant date/time library out there, but the API is really nice.

**Comment by johnclary** [comment]
thanks y'all!

**Pull request merged**